sdk/storage/Azure.Storage.Common/src/AsyncCollection.cs

```diff
@@ -0,0 +1,207 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for
+// license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Storage;
+
+// TODO: Move these types into the Azure.Core package next release?
+
+namespace Azure
+{
+    /// <summary>
+    /// A collection of values that may take multiple service requests to
+    /// iterate over.
+    /// </summary>
+    /// <typeparam name="T">The type of the values.</typeparam>
+    public abstract class AsyncCollection<T> : IAsyncEnumerable<Response<T>>, IEnumerable<Response<T>>
```

Comment:
I have a naming comment further up.  Dropping this note here for relevancy.

I think we need to avoid using *Collection in the name.  This is a sequence.  It's a auto-paging sequence, so I think IAsyncEnumerableWithPaging might be a good option.

Reply:
Collection is actually the recommendation for IEnumerable per https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1710-identifiers-should-have-correct-suffix?view=vs-2019

Reply:
Huh.. I pretty much never see that in practice.

Fair enough.